File: src/main/java/star/team/chat/service/internal/ChatRedisService.java
package star.team.chat.service.internal;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.Cursor;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.ScanOptions;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class ChatRedisService {
    private final RedisTemplate<String, LocalDateTime> localDateTimeRedisTemplate;

    public void markAsRead(Long teamId, Long memberId, LocalDateTime readTime) {
        String key = buildReadKey(teamId, memberId);

        localDateTimeRedisTemplate.opsForValue().set(key, readTime);
    }

    public LocalDateTime getLastReadTime(Long teamId, Long memberId) {
        String key = buildReadKey(teamId, memberId);
        LocalDateTime value = localDateTimeRedisTemplate.opsForValue().get(key);

        return value != null ? value : LocalDateTime.MIN;
    }


    public void deleteAllByTeamId(Long teamId) {
        String pattern = "read:chat:" + teamId + ":*";
        List<String> keysToDelete = new ArrayList<>();
        ScanOptions scanOptions = ScanOptions.scanOptions()
                .match(pattern).count(1000).build();

        try (Cursor<String> cursor = localDateTimeRedisTemplate.scan(scanOptions)) {
            cursor.forEachRemaining(keysToDelete::add);
        }

        if (!keysToDelete.isEmpty()) {
            localDateTimeRedisTemplate.delete(keysToDelete);
        }
    }


    private String buildReadKey(Long teamId, Long memberId) {
        return "read:chat:" + teamId + ":" + memberId;
    }


    private String buildChatKey(Long teamId) {
        return "chat:room:" + teamId;
    }
}

File: src/main/kotlin/star/team/chat/redis/service/ChatRedisService.kt
package star.team.chat.redis.service

import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import star.team.chat.redis.model.entity.ChatLastReadTime
import star.team.chat.redis.repository.ChatLastReadTimeRepository
import java.time.LocalDateTime

@Service
class ChatRedisService(
    private val repository: ChatLastReadTimeRepository
) {

    fun markAsRead(teamId: Long, memberId: Long, readTime: LocalDateTime) {
        val key = generateId(teamId, memberId)
        val entity = repository.findByIdOrNull(key)
            ?.apply { this.lastReadAt = readTime }
            ?: ChatLastReadTime(key, teamId, memberId, readTime)

        repository.save(entity)
    }

    fun getLastReadTime(teamId: Long, memberId: Long): LocalDateTime {
        val key = generateId(teamId, memberId)

        return repository.findByIdOrNull(key)
            ?.lastReadAt
            ?: LocalDateTime.of(1970, 1, 1, 0, 0)
    }

    fun deleteAllByTeamId(teamId: Long) {
        repository.deleteAllByTeamId(teamId)
    }

    private fun generateId(teamId: Long, memberId: Long): String {
        return "$teamId:$memberId"
    }

}